core/src/replay_stage.rs

```diff
@@ -1173,8 +1173,7 @@ impl ReplayStage {
                         vote_account,
                     )
                 } else {
-                    error!("Failed to load tower for {}: {}", node_pubkey, err);
-                    std::process::exit(1);
+                    panic!("Failed to load tower for {}: {}", node_pubkey, err);
```

Comment:
In general, I think it is best if we can avoid explicit `panic!()` / `exit(1)`.

It looks like both callers of `ReplayStage::load_tower()` are in the main `ReplayStage` loop. What if we adjusted this function to return a `Result`, match the `Result` in callers, and set exit flag / break from the loop if the result is an `Err(...)`. IMO, this is the cleanest way to teardown the process

Reply:
ah didn't know we had that custom panic hook, good to know. I agree i think setting the exit flag is the cleanest way, will rework.

Reply:
actually I found that we have a `Finalizer` in the thread that will update the exit flag automatically.

https://github.com/solana-labs/solana/blob/6d154871830df1c87e1cfcac2682251e644d299a/core/src/replay_stage.rs#L576-L578

https://github.com/solana-labs/solana/blob/6d154871830df1c87e1cfcac2682251e644d299a/core/src/replay_stage.rs#L179-L196

This means the panic will allow us to tear down the replay thread which drops the `Finalizer`, sending the exit flag to notify the remaining threads. 
However with the custom panic hook we call `std::process::exit(1)` which could cause a race if other threads have not reacted to the exit flag yet 🤔

Reply:
Oh, I missed the `Finalizer` on first pass, but agreed, seems we have competing teardown.

Regardless, I think moving away from the panics/exit in favor of handling a `Result` is still the right solution. If we break out of the loop in a controlled manner (regardless of whether we set the bool manualy or let the `Finalizer` set it for us), we'll avoid the race with the `exit(1)` in the panic hook.

For the sake of this PR, I think it would be fine to limit cleanup to just the callers of `load_tower()`, and we can cleanup the others panics / exits in subsequent PR's. The other ones look to be buried a couple functions deep so will be slightly more involved

Reply:
agreed, i changed it to a `Result` and return from the loop in order for `Finalizer` to set the flag.

In terms of further cleanup it seems that we should not explicitly `exit(1)` in the panic hook? if we enforce that every thread uses the `Finalizer` then we could avoid the teardown race. 

Where possible we should explicitly set the exit flag or return from thread, but I can imagine more complex scenarios that are not as easy to refactor or unknown panics in which we would still want to shutdown gracefully.

Reply:
> agreed, i changed it to a Result and return from the loop in order for Finalizer to set the flag.

🎉 

> In terms of further cleanup it seems that we should not explicitly exit(1) in the panic hook? if we enforce that every thread uses the Finalizer then we could avoid the teardown race.

Yeah, some refactor across the board might be good so that things are uniform.

> Where possible we should explicitly set the exit flag or return from thread,

Agreed. Along similar lines, it would be nice to have uniform "Service starting" / "Service stopping" log lines ... looked at it for a little once but bailed before finishing

> but I can imagine more complex scenarios that are not as easy to refactor or unknown panics in which we would still want to shutdown gracefully.

Yeah, exactly. At some point somewhere, we need some fail-safe to bail incase we have some thread that doesn't 
have access to the exit flag / is dead-locked / etc. Would be cool if we had a loop that checked if there were any alive threads, and would either report the ones that are still alive or exit if they're all finished. Maybe this loop times out after 15 seconds, but if we could report which threads are alive, it would make it trivial to figure out which threads are not respecting the exit flag and then respond accordingly

> If everyone uses the Finalizer it seems like this should be possible.

Yeah, formalizing some type for widespread use (outside of this PR) could be useful / make things less error prone